arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -726,56 +791,115 @@ void RocksDBMetaCollection::applyUpdates(rocksdb::SequenceNumber commitSeq) {
+      // without holding the lock on these containers though, because of std::multimap's
+      // iterator invalidation rules:
+      // - emplace / insert: No iterators or references are invalidated.
+      // - erase: References and iterators to the erased elements are invalidated. Other references and iterators are not affected.
+      
+      // check for inserts
+      if (applyInserts) {
+        TRI_ASSERT(!applyRemovals);
+      
+        // release the mutex while we modify the tree. this is safe (see above)
+        guard.unlock();
 
-        // check for inserts
-        if (applyInserts) {
-          std::unique_lock<std::mutex> guard(_revisionBufferLock);
-          inserts = std::move(insertIt->second);
-          insertIt = _revisionInsertBuffers.erase(insertIt);
+        TRI_IF_FAILURE("RevisionTree::applyInserts") {
+          THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
         }
-        // check for removals
-        if (applyRemovals) {
-          std::unique_lock<std::mutex> guard(_revisionBufferLock);
-          removals = std::move(removeIt->second);
-          removeIt = _revisionRemovalBuffers.erase(removeIt);
+
+        // apply inserts, without holding the lock
+        // if this throws we will not have modified _revisionInsertBuffers
+        try {
+          _revisionTree->insert(insertIt->second);
+        } catch (std::exception const& ex) {
+          // it is pretty bad if this fails, so we want to see it in our tests
+          // and not overlook it.
+          LOG_TOPIC("27811", ERR, Logger::ENGINES)
+              << "unable to apply revision tree updates for " 
+              << _logicalCollection.vocbase().name() << "/" << _logicalCollection.name() 
+              << ": " << ex.what();
+          TRI_ASSERT(false);
+          throw;
         }
-      }
 
-      // no inserts or removals left to apply, drop out of loop
-      if (inserts.empty() && removals.empty()) {
-        break;
+        // move iterator forward, we need the mutex for this
+        guard.lock();
+        insertIt = _revisionInsertBuffers.erase(insertIt);
       }
 
-      // apply inserts
-      if (!inserts.empty()) {
-        _revisionTree->insert(inserts);
-        inserts.clear();
-      }
+      // check for removals
+      if (applyRemovals) {
+        TRI_ASSERT(!applyInserts);
+        
+        // release the mutex while we modify the tree. this is safe (see above)
+        guard.unlock();
+        
+        TRI_IF_FAILURE("RevisionTree::applyRemoves") {
+          THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+        }
 
-      // apply removals
-      if (!removals.empty()) {
-        _revisionTree->remove(removals);
-        removals.clear();
+        // apply removals, without holding the lock
+        // if this throws we will not have modified _revisionRemovalBuffers
+        try {
+          _revisionTree->remove(removeIt->second);
+        } catch (...) {
+          // this should never fail
+          TRI_ASSERT(false);
```

Comment:
Ok, so this _should_ not fail, but shouldn't we at least write a log if it actually does?